File: src/main/java/me/itzisonn_/meazy/parser/ast/Statement.java
package me.itzisonn_.meazy.parser.ast;

import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.runtime.environment.Environment;
import org.jspecify.annotations.NullMarked;

/**
 * Represents unit of the program with multiple lines possible
 *
 * @see Expression
 */
@NullMarked
public interface Statement { //TODO JAVADOC
    void emit(InstructionsSet instructionsSet, Environment environment, Statement parent);
}

File: src/main/kotlin/me/itzisonn_/meazy/parser/ast/statement/Statement.kt
package me.itzisonn_.meazy.parser.ast.statement

import me.itzisonn_.meazy.parser.ast.ParentMap
import me.itzisonn_.meazy.parser.ast.ProgramUnit
import me.itzisonn_.meazy.parser.ast.SymbolMap
import me.itzisonn_.meazy.runtime.data.symbol.Symbol
import me.itzisonn_.meazy.runtime.environment.Environment

/**
 * Represents program unit with multiple lines possible
 */
interface Statement : ProgramUnit

interface LocalStatement : Statement {
    fun alwaysReturns(): Boolean
}

interface DeclarationStatement<T : Symbol> : Statement {
    context(parents: ParentMap, symbols: SymbolMap)
    fun declare(environment: Environment)

    context(parents: ParentMap, symbols: SymbolMap)
    fun resolve(environment: Environment)
}